src/libraries/Microsoft.Extensions.Http/src/DependencyInjection/HttpClientBuilderExtensions.cs

```diff
@@ -264,6 +265,72 @@ public static IHttpClientBuilder ConfigureHttpMessageHandlerBuilder(this IHttpCl
             return builder;
         }
 
+#if NET5_0_OR_GREATER
+        /// <summary>
+        /// Adds or updates <see cref="SocketsHttpHandler"/> as a primary handler for a named <see cref="HttpClient"/>. If provided,
+        /// also adds a delegate that will be used to configure the primary <see cref="SocketsHttpHandler"/>.
+        /// </summary>
+        /// <param name="builder">The <see cref="IHttpClientBuilder"/>.</param>
+        /// <param name="configureHandler">Optional delegate that is used to configure the primary <see cref="SocketsHttpHandler"/>.</param>
+        /// <returns>An <see cref="IHttpClientBuilder"/> that can be used to configure the client.</returns>
+        /// <remarks>
+        /// <para>
+        /// If a primary handler was already set to be <see cref="SocketsHttpHandler"/> by previously calling, for example,
+        /// <see cref="ConfigurePrimaryHttpMessageHandler(IHttpClientBuilder, Func{HttpMessageHandler})"/> or
+        /// <see cref="UseSocketsHttpHandler(IHttpClientBuilder, Action{ISocketsHttpHandlerBuilder})"/>, then the passed <paramref name="configureHandler"/>
+        /// delegate will be applied to the existing instance. Otherwise, a new instance of <see cref="SocketsHttpHandler"/> will be created.
+        /// </para>
+        /// </remarks>
+        [UnsupportedOSPlatform("browser")]
+        public static IHttpClientBuilder UseSocketsHttpHandler(this IHttpClientBuilder builder, Action<SocketsHttpHandler, IServiceProvider>? configureHandler = null)
```

Comment:
Why is it a Use vs Add?

Reply:
Because then it would have been ConfigurePrimaryHandlerToBeSocketsHttpHandler 😅

Point is, you don't have SocketsHttpHandler in HttpClientFactory by default - you have some PrimaryHandler, so if you'd say ConfigureSocketsHttpHandler, the question is -- where did it come from, so that we're configuring it?

So to me, saying UseSocketsHttpHandler was much clearer in explaining what happens - after you call it, the factory would be using SocketsHttpHandler in this named client

Reply:
While I agree that "Use" is very rarely used, I don't think that `ConfigureSocketsHttpHandler` is better, for the reason I described above

> when it's calling a Configure* method

It's calling Configure**Primary**HttpMessageHandler method. "Primary" is an important piece of information in relation to Configure, as Primary handler is part of HttpClientFactory's "world", it's a "property" on an HttpClient's configuration.

Primary handler property is a "left-hand operand" in the assignment, and SocketsHttpHandler value is a "right-hand operand", so to say.

But that's my opinion; I've sent an email to the API review board to discuss this further.

Reply:
Given the example usage, I'm not sure I like "Configure..." given the nested calls to `Configure(` in the builder callback.

```csharp
services.AddHttpClient("baz")
    .ConfigureSocketsHttpHandler(builder =>
        builder.Configure(configuration.GetSection("HttpClientSettings:baz"))
               .Configure((handler, _) => handler.SslOptions = new SslClientAuthenticationOptions
               {
                   RemoteCertificateValidationCallback = delegate { return true; },
               })
    );
```

@davidfowl During API review, I thought your feedback would be to have it return the `ISocketsHttpHandlerBuilder` to reduce nesting. But if we did that, "Use..." would feel more out of place. "Add..." would be far more conventional given things like `AddAuthentication` and `AddMvc`, but it is a little weird that add really means replace in this case. But for completeness, the builder-returning API would look something like this:

```csharp
services.AddHttpClient("baz")
    .AddSocketsHttpHandler()
    .Configure(configuration.GetSection("HttpClientSettings:baz"))
    .Configure((handler, _) => handler.SslOptions = new SslClientAuthenticationOptions
    {
        RemoteCertificateValidationCallback = delegate { return true; },
    });
```

But as @CarnaViire pointed out in API review, this could be annoying if you need the `IHttpClientBuilder` for anything else, you'd need to store the builder in a variable unless `AddSocketsHttpHandler` is the last part of the chain, so we decided against it.

```csharp
var httpClientBuilder = services.AddHttpClient("baz");

httpClientBuilder.AddSocketsHttpHandler()
    .Configure(configuration.GetSection("HttpClientSettings:baz"))
    .Configure((handler, _) => handler.SslOptions = new SslClientAuthenticationOptions
    {
        RemoteCertificateValidationCallback = delegate { return true; },
    });

httpClientBuilder.SomethingElseThatReturnsABuilder()
    .DoSome()
    .MoreThings();
```

Reply:
To get a sense of why Use over Configure here, what else would we call Use* on this type?

Reply:
Potentially something related to a specific primary handler? e.g. `UseWinHttpHandler(...)` -> + configuration specific to WinHttpHandler

[Scope mismatch](https://github.com/dotnet/runtime/issues/47091) fix API could be named `UseExistingScope(true)` or `UseHandlerScope(false)`

Something like `UseMinimalLogging(...)` that I've described as a possible later expansion in https://github.com/dotnet/runtime/issues/77312, to highlight it will *replace* existing logging, as opposed to `Add...Logging`, that only adds and you have to call `RemoveAllLoggers` beforehand

Reply:
Today we have Set* for that right?  Here's what I see today:

```C#
builder.Services.AddHttpClient("Custom")
    .RedactLoggedHeaders(header => header.Equals("Authenticaton", StringComparison.OrdinalIgnoreCase))
    .ConfigureHttpClient(client => client.Timeout = TimeSpan.FromSeconds(10))
    .ConfigurePrimaryHttpMessageHandler(() => new HttpClientHandler()
    {
        AutomaticDecompression = System.Net.DecompressionMethods.GZip
    })
    .SetHandlerLifetime(TimeSpan.FromMinutes(5))
    .AddHttpMessageHandler<AuthenticationHander>();
```

You're suggesting we replace Set with Use because we're setting a property that is being configured (even that that's all configure* methods do right?), but because there's no arguments (or because it's optional).

After this change, it'll look like this:

```C#
builder.Services.AddHttpClient("Custom")
    .RedactLoggedHeaders(header => header.Equals("Authenticaton", StringComparison.OrdinalIgnoreCase))
    .ConfigureHttpClient(client => client.Timeout = TimeSpan.FromSeconds(10))
    .UseSocketsHttpHandler(h =>
    {
        h.PooledConnectionLifetime = TimeSpan.FromMinutes(5);
    })
    .SetHandlerLifetime(TimeSpan.FromMinutes(5))
    .AddHttpMessageHandler<AuthenticationHander>();
```

Reply:
> Today we have Set* for that right?

"obj.Set**Property**(value)" translates into "obj.**Property** = value", doesn't it? If we choose Set* for the API, you would get "obj.**value???** = ???" instead.... It's the same logical error IMHO as with Configure*

Reply:
I'm not married to Use* in particular, I just believe Set*, Configure* and Add* are all logically wrong here.....

Reply:
BTW doesn't ASP.NET have `UseHttpSys(...)`, `UseIIS(...)`, `UseKestrel(...)`, `UseQuic(...)` for exactly same core meaning?

Reply:
Use(this IWebHostBuilder) - Adding a combination of services, logging and configuration
Use(this IApplicationBuilder) - Adding middleware to the pipeline
Map(this IEndpointRouteBuilder) - Adding a route
Add(this IServiceCollection) - Adding services to the DI container
Configure<T>(this IServiceCollecton) - Configuring options for T

Reply:
@CarnaViire can you open a follow up issue?

Reply:
@davidfowl https://github.com/dotnet/runtime/issues/89023#issuecomment-1638468865